lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -205,6 +205,155 @@ static Node *simplifyNode(Node *node, Function *F) {
   return node;
 }
 
+// 2D vector of dimensions mapping a dimension that is reshaped into either 1
+// or 2 dimensions by the first reshape node in a ChannelShuffle operation
+using ChannelShuffleDimGroups =
+    llvm::SmallVector<llvm::SmallVector<unsigned_t, 2>, max_tensor_dimensions>;
+
+/// fills \p dimGroups with the groups of dimensions of the result of \p node
+/// that get separated. This grouping is only possible if \p node is the first
+/// Reshape node in a ChannelShuffle operation (Reshape->Tranpose->Reshape)
+/// \returns true if successfully grouped output dimensions and false otherwise
+static bool
+groupDimensionsFromChannelShuffleReshape(const ReshapeNode &node,
+                                         ChannelShuffleDimGroups &dimGroups) {
+  const auto dims = node.getDims();
+  const auto inputDims = node.getInput().dims();
+  if (dims.size() != inputDims.size() + 1) {
+    return false;
+  }
+  dimGroups.resize(inputDims.size());
+  bool foundReshapedDim = false;
+  for (size_t dimIdx = 0, groupIdx = 0; dimIdx < dims.size();
+       ++dimIdx, ++groupIdx) {
+    auto &group = dimGroups[groupIdx];
+    group.push_back(dimIdx);
+    if (std::find(inputDims.begin(), inputDims.end(), dims[dimIdx]) ==
+        inputDims.end()) {
+      // make sure only one input dimension is changed and there's at least one
+      // more dim
+      if (foundReshapedDim || dimIdx >= dims.size() - 1) {
+        return false;
+      }
+      group.push_back(++dimIdx);
+      foundReshapedDim = true;
+    }
+  }
+  return true;
+}
+
+// Sink Transpose below ChannelShuffle node sequence. For example
+// (Transpose_1->Reshape_1->Transpose_2->Reshape_2) becomes
+// (Reshape_1->Transpose_2->Reshape_2->Transpose_1) and Reshape_1,
+// Transpose_2, Reshape_2 are reconstructed so that the end result is the
+// ChannelShuffle operation on the same dimention as before Transpose_1 was
+// sunk.
+static bool sinkTranposeBelowChannelShuffle(Function *F, Node *node) {
```

Comment:
This function is long. Can we refactor some of the logic in the bottom of the function to new functions?

Reply:
I agree it's kinda long but I don't see too many opportunities for extracting something cleanly from this in its current form, do you have anything in particular that you'd like to see abstracted? Two thoughts I have on changing it though are
1. I could rewrite the bottom part using `Graph::createChannelShuffle` and then change the first portion of the code to deduce the correct parameters to the original ChannelShuffle and then and apply the inverse transpose to those to get the new parameters to feed to `createChannelShuffle`
2. we could go ahead and create a ChannelShuffle node as @jfix71 mentioned earlier that get's created at graph load time instead of immediately forming the reshape->transpose->reshape and do optimizations on that node however the shufflenet onnx file has this reshape->transpose->reshape already in it so in cases like this we'd have to first collapse this pattern to a ChannelShuffle when it's seen in a graph and then lower it back to reshape->transpose->reshape after the first round of graph optimization.
If we think that either of those is better than the current approach I can do that instead.

Reply:
I personally think that option 1 is a good middle ground here to go with

Reply:
Yeah, I think we should not add a ChannelShuffleNode and just try to refactor the optimization here to make it as readable as possible.

Reply:
b5f6a7b does this refactoring and I think it ended up being much cleaner

Reply:
Something that's not good about having this in another function is that later on if someone add an if branch to `sinkCode()` for `ReshapeNode` in the same style as the other branches there (with a `continue` at the bottom then this code could potentially get skipped always.

Reply:
Agreed -- maybe you could change it so that you check if the node is a `ReshapeNode` in `sinkCode()`, and then pass the ReshapeNode into `sinkTranposeBelowChannelShuffle()`.

Reply:
@jfix71 good suggestion